### PR TITLE
GH-1874: Fix issue of invalid null-null-deltas due to resource-creation, build-cancellation, and resource-deletion

### LINKED ISSUE #1874: Fix issue of invalid null-null-deltas due to resource-creation, build-cancellation, and resource-deletion
The following stack trace happens when a resource is 
(1) created,
(2) a cancellation happens, and
(3) the resource is removed again.
What should not happen is that the build crashes during this course of actions. A crash could happen since the delta that is created at (1) will be merged with the delta at (3). This merge will result in a delta from old==null to new==null, which is an invalid delta.

```
ERROR - error during queued task: 
java.lang.AssertionError: 'old!=_new' constraint violated
	at org.eclipse.xtext.resource.impl.DefaultResourceDescriptionDelta.<init>(DefaultResourceDescriptionDelta.java:30)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.mergeWithUnreportedDeltas(XWorkspaceBuilder.java:438)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.recordBuildProgress(XWorkspaceBuilder.java:420)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.doIncrementalBuild(XWorkspaceBuilder.java:362)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.lambda$5(XWorkspaceBuilder.java:281)
	at org.eclipse.n4js.ide.xtext.server.build.BuilderFrontend.lambda$5(BuilderFrontend.java:157)
	at org.eclipse.n4js.ide.server.concurrent.N4JSQueuedExecutorService.lambda$0(N4JSQueuedExecutorService.java:35)
	at org.eclipse.n4js.ide.xtext.server.QueuedExecutorService$QueuedTask.run(QueuedExecutorService.java:119)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by mmews-n4**
closes #1874

### CODE REVIEW

File: plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java [hunk 1]
```diff
@@ -435,7 +435,9 @@ private void mergeWithUnreportedDeltas(List<IResourceDescription.Delta> newDelta
 					toBeConsideredDeltas.remove(unreportedDelta);
 					IResourceDescription _old = unreportedDelta.getOld();
 					IResourceDescription _new = newDelta.getNew();
-					toBeConsideredDeltas.add(new DefaultResourceDescriptionDelta(_old, _new));
+					if (_old != null || _new != null) {
+						toBeConsideredDeltas.add(new DefaultResourceDescriptionDelta(_old, _new));
+					}
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java:440]: as discussed: consider keeping the newDelta if `_old == _new == null` to have dependent resources to be built

File: tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/server/AbstractIdeTest.java [hunk 2]
```diff
@@ -169,6 +172,7 @@ static final public void resetPrintStreams() {
 	/** Deletes the test project in case it exists. */
 	@After
 	final public void deleteTestProject() {
+		assertNoErrorsInLog();
```
Comment (mor-n4) [inline tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/server/AbstractIdeTest.java:175]: Great! :tada:

File: tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend [hunk 3]
```diff
@@ -27,31 +27,48 @@ import org.eclipse.xtext.util.Strings
...
 
+	private static final AtomicReference<URI> cancelOnResource = new AtomicReference(null);
+	private static final AtomicReference<URI> cancelAfterResource = new AtomicReference(null);
+	
 	/* 
 	 * Iff greater zero, then cancel before building the n-th resource. Cancellation happens before the very first 
 	 * resource with a cancelCounter set to 1.
 	 */
-	private static final AtomicInteger cancelCounter = new AtomicInteger(0);
+	private static final AtomicInteger cancelOnCount = new AtomicInteger(0);
```
Comment (mor-n4) [inline tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend:45]: Good idea to rename this. But what about `cancelWhenZero`?

File: tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend [hunk 4]
```diff
@@ -219,4 +236,99 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest
 		assertEquals("(Error, [1:24 - 1:26], Couldn't resolve reference to IdentifiableElement 'a3'.)",
 			issues.sort.join(Strings.newLine))
 	}
+
+	/**
+	 * This test case tests a very specific edge case: A resource is 
+	 * (1) created,
+	 * (2) a cancellation happens, and
+	 * (3) the resource is removed again.
```
Comment (mor-n4) [inline tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend:244]: maybe: "... is removed again _before the next incremental build iteration_." ? (or something similar)

**Review by mor-n4 — APPROVED** [review]
:+1:
